spec.md

```diff
@@ -2374,6 +2407,36 @@ SInt("h-2A")
 SInt("h-2a")
 ```
 
+## Enum Expressions
+
+An enumeration can be constructed by specifying the variant tag, an optional
+data value expression, and a return type specifier.
+
+``` firrtl
+#a : [#a, #b, #c]
```

Comment:
In human language, does this mean "a thing #a is of the anonymous enum type that might be either #a, #b, #c".

It seems that the most common cases for enums isn't covered by this PR, of, (to give the classic example) "A thing Add, that may be of Add (represented as 0) , Sub (represnted as 1), Mul(represented as 2), .... "

Is there a reason we are omitting that 'common' case and then adding Some of a totally unknown value ...?

I guess I'd call this PR "union" instead of "enum" in its current form...?

Reply:
Yeah these correspond to discriminated unions.  The name `enum` comes from what it means in [Rust](https://doc.rust-lang.org/reference/items/enumerations.html) and [Swift](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/enumerations/). The idea is that if every variant of an enum has no data (i.e. the data is a `UInt<0>`), then we will lower this to an actual enum.  If there is a data "payload", then it will be lowered to a tagged union, something like
```verilog
struct {
  enum tag {a, b, c}
  union {
    int a_data,
    int b_data,
    int c_data
  } data
}
```

Reply:
would you be willing to write an "opcode" like example, that has the same aha I see what you are doing here as you are doing with `some` and `none`"? Sorry I'm not getting it :)

Reply:
i mean, it's not like a_data, b_data, c_data can be any number.

Reply:
A simple enum example:
```
module Executor:
  in opcode : [#add, #sub, #mul]
  in arg0: UInt<8>
  in arg1 : UInt<8>
  out result : UInt<8>
  match opcode:
    #add:
      result <= add(arg0, arg1)
    #sub:
      result <= sub(arg0, arg1)
    #mul:
      result <= mul(arg0, arg1)
```

Reply:
An enum with a payload:
```
module Foo:
  in signal : { data : [#valid: UInt<8>, #invalid], flip ready: UInt<1>}
  output o : UInt<8>

  signal.ready <= UInt<1>(1)
 
  match signal.data:
    #valid(x):
      o <= x
    #invalid:
      o <= UInt<8>(0)
```

Reply:
thanks! great and clarifying examples. not sure if it is too many to add to the spec.